File: ios/SecretaryVaultAccess/Sources/SecretaryVaultAccessTesting/FakeFolderWatch.swift
import Foundation
import SecretaryVaultAccess

/// In-memory `FolderWatchPort`. Tests call `emit` to deliver pulses on demand.
public final class FakeFolderWatch: FolderWatchPort {
    public private(set) var started = false
    public private(set) var stopCount = 0
    public private(set) var startCount = 0
    /// Set before `start` to make it throw.
    public var startError: Error?
    private var onPulse: (@MainActor (MonotonicInstant) -> Void)?

    public init() {}

    public func start(onPulse: @escaping @MainActor (MonotonicInstant) -> Void) throws {
        if let startError { throw startError }
        startCount += 1
        started = true
        self.onPulse = onPulse
    }

    public func stop() {
        stopCount += 1
        started = false
        onPulse = nil
    }

    /// Test hook: deliver a pulse to the registered callback.
    @MainActor public func emit(at instant: MonotonicInstant) { onPulse?(instant) }
}

File: android/vault-access/src/test/kotlin/org/secretary/sync/FakeFolderWatch.kt
package org.secretary.sync

/**
 * In-memory [FolderWatchPort] for host tests. The test drives raw pulses via [emit].
 * [startError], if set, is thrown by [start] (before any state changes) to exercise the
 * monitor's start-failure roll-back path.
 */
class FakeFolderWatch : FolderWatchPort {
    var startCount: Int = 0
        private set
    var stopCount: Int = 0
        private set
    var startError: Throwable? = null
    private var onPulse: ((MonotonicInstant) -> Unit)? = null

    val started: Boolean get() = onPulse != null

    override fun start(onPulse: (MonotonicInstant) -> Unit) {
        startError?.let { throw it }
        this.onPulse = onPulse
        startCount++
    }

    override fun stop() {
        onPulse = null
        stopCount++
    }

    /** Deliver a pulse through the registered callback (simulates an OS event). */
    fun emit(at: MonotonicInstant) {
        val cb = onPulse ?: error("emit called while watch not started")
        cb(at)
    }
}
